pkg/agent/pipeline.go

```diff
@@ -0,0 +1,203 @@
+	// Module at index 1 should always be first meta module and already be set
+	gate := iPort.Connect(s.Modules[1], true, nil)
+	log.Debugf("Port %s via gate: %d connected to module %s", iPort.Name, gate, s.Modules[1].GetName())
+
+	// Update forwarder module with reverse linkage
+	gate = s.Modules[1].Connect(iPort, false, nil)
+	log.Debugf("Module %s via gate: %d connected to port %s", s.Modules[1].GetName(), gate, iPort.Name)
+
+	// Create Egress port
+	ePortMod := network.Module{Name: fmt.Sprintf("%s_egress", name),
+		IGates: make(map[network.Gate]network.PipelineModule)}
+	ePort := &network.EgressPort{Port: port, Module: ePortMod}
+	return ePort, nil
+}
+
+func getForwarderModule(forwarder string, metaKey string) (network.PipelineModule, error) {
+	switch forwarder {
+	case L2DriverMode:
+		mod := network.Module{Name: fmt.Sprintf("Switch_%s", metaKey),
+			IGates: make(map[network.Gate]network.PipelineModule),
+			EGates: make(map[network.Gate]network.PipelineModule),
+		}
+		sMod := &network.Switch{Module: mod, L2FIB: make(map[string]network.Gate)}
+		return sMod, nil
+	default:
+		return nil, fmt.Errorf("invalid forwarder given: %s", forwarder)
+	}
+}
+
+// Init is responsible for initializing a new pipeline.
+// It checks the abstract pipeline for a network and builds the pipeline
+// for this port. If meta pipeline is not nil, s is a port pipeline.
+func (s *NimbessPipeline) Init(port string, forwarder string, metaPipeline *NimbessPipeline) error {
+	if metaPipeline != nil {
+		log.Infof("Initializing pipeline for port %s", port)
+		if len(metaPipeline.Modules) == 0 {
+			return fmt.Errorf("empty Module list for metapipeline: %v", metaPipeline)
+		}
+		// first index in port pipeline slice is always ingress port, reserve it
+		s.Modules = append(s.Modules, nil)
+		// Need to do deep copy Meta pipeline for port, and make unique module names
+		// Update linked list of modules by making new modules from meta pipeline for this pipeline
+		s.createPipelineLinkedList(metaPipeline.Modules[0], port)
+	} else {
+		if len(s.Modules) > 0 {
+			log.Debugf("Pipeline already exists, ignoring Init on meta pipeline: %v", metaPipeline)
+			return nil
+		}
+		// Need to create meta pipeline from scratch, just include forwarder module
+		mod, err := getForwarderModule(forwarder, s.MetaKey)
+		if err != nil {
+			return err
+		}
+		// Set meta module name to be itself, just in case it gets copied by port pipeline module
+		mod.SetMeta(mod.GetName())
+		s.Modules = append(s.Modules, mod)
+		log.Debugf("Added forwarder module to meta pipeline: %s", mod.GetName())
+	}
+
+	return nil
+}
+
+// createPipelineLinkedList recursively creates modules and then updates its linkage from meta pipeline module
+// module should be the a pointer to the first module in a meta pipeline
+// port is name of new port used for this new port pipeline
+func (s *NimbessPipeline) createPipelineLinkedList(module network.PipelineModule, port string) network.PipelineModule {
```

Comment:
seems like s smell to have linkedlist in the function name

Reply:
huh?

Reply:
Having linkedList in the function name implies the implementation, so just remove that part.

Reply:
ok will remove the type implication